ui/src/main/java/edu/wpi/grip/ui/pipeline/StepController.java

```diff
@@ -0,0 +1,120 @@
+    private final OutputSocketController.Factory outputSocketControllerFactory;
+    private final Step step;
+
+    private NodeControllerObservableListMap<InputSocketController, Node> inputSocketMapManager;
+    private NodeControllerObservableListMap<OutputSocketController, Node> outputSocketMapManager;
+
+    /**
+     * Used for assisted injects.  Guice will automatically create an instance of this interface so we can create
+     * step controllers.  This lets us use injection with StepController even though it requires a {@link Step}
+     * (which is not an injected dependency).
+     */
+    public interface Factory {
+        StepController create(Step step);
+    }
+
+    @Inject
+    StepController(EventBus eventBus, InputSocketControllerFactory inputSocketControllerFactory, OutputSocketController.Factory outputSocketControllerFactory, @Assisted Step step) {
+        this.eventBus = eventBus;
+        this.inputSocketControllerFactory = inputSocketControllerFactory;
+        this.outputSocketControllerFactory = outputSocketControllerFactory;
+        this.step = step;
+    }
+
+    public void initialize() {
+        inputSocketMapManager = new NodeControllerObservableListMap<>(inputs.getChildren());
+        outputSocketMapManager = new NodeControllerObservableListMap<>(outputs.getChildren());
+
+        root.getStyleClass().add(StyleClassNameUtility.classNameFor(step));
+        title.setText(step.getOperation().getName());
+        step.getOperation().getIcon().ifPresent(icon -> this.icon.setImage(new Image(icon)));
+
+        // Add a SocketControlView for each input socket and output socket
+        for (InputSocket<?> inputSocket : step.getInputSockets()) {
+            inputSocketMapManager.add(inputSocketControllerFactory.create(inputSocket));
+        }
+
+        for (OutputSocket<?> outputSocket : step.getOutputSockets()) {
+            outputSocketMapManager.add(outputSocketControllerFactory.create(outputSocket));
+        }
+    }
+
+    /**
+     * @return An unmodifiable collection of {@link InputSocketController}s corresponding to the input sockets of this step
+     */
+    @SuppressWarnings("unchecked")
+    public Collection<InputSocketController> getInputSockets() {
+        return inputSocketMapManager.keySet();
+    }
+
+    /**
+     * @return An unmodifiable collection of {@link InputSocketController}s corresponding to the output sockets of this step
+     */
+    @SuppressWarnings("unchecked")
+    public Collection<OutputSocketController> getOutputSockets() {
+        return outputSocketMapManager.keySet();
+    }
+
+    public VBox getRoot() {
+        return root;
+    }
+
+    public Step getStep() {
+        return step;
+    }
+
+    @FXML
+    private void deleteStep() {
+        eventBus.post(new StepRemovedEvent(step));
+    }
+
+    @FXML
+    private void moveStepLeft() {
+        eventBus.post(new StepMovedEvent(step, -1));
+    }
+
+    @FXML
+    private void moveStepRight() {
```

Comment:
Ohhhh. `@FXML` methods can be private! This should be done more. Or made private.